madanalysis/multiweight/histogram.py

```diff
@@ -0,0 +1,552 @@
+    def all_scaled_weights(self):
+        return [
+            (pos - abs(neg)) * self.scale
+            for pos, neg in zip(self._positive_weights, self._negative_weights)
+        ]
+
+    @property
+    def has_scale_unc(self) -> bool:
+        """Does the histogram has scale unc"""
+        return self.weight_collection.has_scale
+
+    @property
+    def scale_uncertainties(self) -> Tuple[np.ndarray, np.ndarray]:
+        """
+        Retreive scale uncertainties
+
+        Returns:
+            ``Tuple[np.ndarray, np.ndarray]``:
+            lower and upper uncertainties per bin
+        """
+        central_weights = self.weights * self.scale
+        bins = self.all_scaled_weights
+
+        if not self.weight_collection.has_scale:
+            central = [b[self.central_idx] for b in bins]
+            return central, central
+
+        weight_loc = (
+            self.weight_collection.get_scale_vars(
+                self.n_point_scale_variation, self.dynamic_scale_choice
+            )
+            .pdfset(self.nominal_weight.pdfset)
+            .loc
+        )
+
+        upper, lower = [], []
+        for current_bin in bins:
+            current_weights = [current_bin[wloc] for wloc in weight_loc]
+            upper.append(max(current_weights))
+            lower.append(min(current_weights))
+        return (
+            np.clip(central_weights - np.array(lower), 0.0, None),
+            np.clip(np.array(upper) - central_weights, 0.0, None),
+        )
+
+    @property
+    def has_pdf_unc(self) -> bool:
+        """Does the histogram has PDFs"""
+        return len(self.weight_collection.pdfsets) > 1
+
+    @property
+    def pdf_uncertainties(self) -> Tuple[np.ndarray, np.ndarray]:
+        """
+        Retreive PDF uncertainties
+
+        Returns:
+            ``Tuple[np.ndarray, np.ndarray]``:
+            lower and upper uncertainties per bin
+        """
+        if not self.has_pdf_unc:
+            return [self.weights * self.scale] * 2
+
+        method = "replicas"  # "eigenvector"
+
+        bins = self.all_scaled_weights
+
+        pdfids = self.weight_collection.pdfsets
+        pdfids.pop(pdfids.index(self.nominal_weight.pdfset))
+
+        pdf_sets = WeightCollection([self.nominal_weight])
+
+        new_bins = []
+
+        # Replicas
+        if method == "replicas":
+            # 2202.13416 eq 3.2
+            pdf_sets += self.weight_collection.pdfset(pdfids)
+            pdf_locs = pdf_sets.loc
+            mean_bin_value = np.array(
+                [np.mean([bn[loc] for loc in pdf_locs]) for bn in bins]
+            )
+
+            for bn, mean_val in zip(bins, mean_bin_value):
+                new_bins.append(
+                    np.sqrt(
+                        sum((bn[loc] - mean_val) ** 2 for loc in pdf_locs)
+                        / (len(pdf_locs) - 1)
+                    )
+                )
+            return new_bins, new_bins
+
+        # Eigenvectors
+        # ! Unclear 2202.13416 eq 3.1
```

Comment:
@BFuks the eigenvectors are not clear to me, not sure how to get +i and -i to compare because sign changes.

Reply:
Paired eigenvectors correspond to pairs of subsequent sets.